### ISSUE TITLE
rust-lang repos that do not declare licenses

### ISSUE THREAD

**Issue by brson** [rust-lang/rust#25664]
It came up today that kate-config does not have a license file. I filed a [PR](https://github.com/rust-lang/kate-config/pull/9) to fix that, but there are others.

* crates.io-index - I'm not sure if this needs a copyright license since it isn't a work of creative expression, but it might still be worth putting something explicit in there just to fend off questions. Maybe there's some sort of 'open-data' license that's appropriate.
* blog.rust-lang.org
* regex - Declared in Cargo.toml but no accompanying license files.
* rfcs - This is a mess. It should have the MIT/ASL license per our other docs but we didn't have the foresight to do it.

**Comment by edunham** [comment]
[ ] crates.io-index - I'm not sure if this needs a copyright license since it isn't a work of creative expression, but it might still be worth putting something explicit in there just to fend off questions. Maybe there's some  sort of 'open-data' license that's appropriate.  https://github.com/rust-lang/crates.io-index/issues/3 
[ ] blog.rust-lang.org  https://github.com/rust-lang/blog.rust-lang.org/issues/63 CC-BY suggested                                                        
[x] regex - Declared in Cargo.toml ~~but no accompanying license files.~~ has dual Apache+MIT        
[ ] rfcs - This is a mess. It should have the MIT/ASL license per our other docs but we didn't have the foresight to do it. https://github.com/rust-lang/rfcs/issues/1259                                 
[x] rust-buildbot - now dual-licensed                                                             
[ ] meeting-minutes - Doesn't matter probably.                                  
[ ] rust-www - This is tricky. It contains a COPYRIGHT file, but I believe we put it there to have a license to link to for Rust itself. We can probably argue that the entire website falls under that license.  https://github.com/rust-lang/rust-www/issues/174                   
[ ] rust.vim https://github.com/rust-lang/rust.vim/issues/45
[ ] rust-mode https://github.com/rust-lang/rust-mode/issues/91 
[x] rust-packaging - contains licenses now
[x] gedit-config - dual-licensed
[x] rustup - rustup.sh itself contains the license, ~~but not the rest.~~ and repo contains both license files
[x] rust-installer - ditto                                   
[ ] rust-wiki-backup - decommissioned partially because it was not properly licensed 
[ ] rust-guidelines - repo was decommissioned and moved 
[x] nano-config - ~~Only contains copyright notice, not license~~ Has both licenses             
[x] zsh-config  - ditto

**Comment by cmr** [comment]
@edunham CC0 seems suitable for the crates.io-index, if anything.

**Comment by eshellman** [comment]
Maybe I missed something, but this repo (rust-lang/rust) is also missing a license file.

**Comment by steveklabnik** [comment]
@eshellman https://github.com/rust-lang/rust/blob/master/LICENSE-APACHE and https://github.com/rust-lang/rust/blob/master/LICENSE-MIT

**Comment by eshellman** [comment]
Not sure how I missed that. not enough coffee yet.

**Comment by sanmai-NL** [comment]
I would like to take this issue up again.

What is the status?
Where is help needed?

Specifically on [licenses repeated in source code files](e.g. [`mk/main.mk`](https://github.com/rust-lang/rust/blob/a005b6785935d7e92e87774c8f585839ddd12e46/mk/main.mk#L1-L9)), I hope we will soon make good improvement across the board. I propose that we remove them or reduce them to at most a single line. 

**Arguments**
1. They hurt readability and conciseness;
2. They are a small extra hurdle to contributing changes (esp. from the GitHub web UI) since it makes you scroll down on every file;
3. Their use across source files violates the do-not-repeat-yourself principle;
4. Probably in large part due to the previous, they are undermaintained. E.g. something as basic as the year is messed up [in a prominent place](https://github.com/rust-lang/rust/blob/7bf54f90d619fc87af4c8b660a80166c40048cb5/src/bootstrap/mk/Makefile.in#L1).

**Issues, my positions on them**
A. Who supports [this position](http://programmers.stackexchange.com/a/125841), that a single central license file in each repo should be used exclusively instead? I do, for sure. Only for GPL and/or mixed licensing it may be required to refer to the GPL again at the top of every file, but if we do then let us please use a one-line comment referring to the license file.

B. Who supports my position that a copyright claim is redundant, and that dating the copyright is even more redundant (certainly as the project keeps record using a VCS), and should be removed everywhere as a maintenance burden?

@brson: What has led to the situation that [multiple licenses are applicable to the `rust-lang` code base](https://github.com/rust-lang/rust#license)? What can still be done to harmonize this to a single license?

**Comment by eshellman** [comment]
The copyright year only matters 95 years after the date of first publication (treating this as a work of corporate authorship). So this repo begins to enter the US public domain in 2098 under current law. Best practice is to maintain copyright notice in a single file and let VCS sweat the details.

**Comment by brson** [comment]
@sanmai-NL the issues you are raising seem to be different from this issue, which is about ensuring that individual rust-lang repos declare a license (and on that basis I think we can close this issue - most of the remaining undeclared repos are not worth the effort). So you might consider opening another issue if you think the Rust source copyright is not handled properly.

That said, the exact handling of the license declarations in this repo has been debated repeatedly over the years and I suspect there isn't much enthusiasm for re-litigating any aspect of it. The current scheme is as recommended by Mozilla legal long ago. It is true though that if we were to do it over with the benefit of hindsight I suspect we would push harder not to have the notice at the top of each file, nor to mention the copyright year. [Here is an old discussion on the copyright year](https://github.com/rust-lang/rust/issues/9070). We decided that mentioning the copyright year is useless, and that they should not be updated.

The primary Rust codebase is Apache licensed because we want a permissive license with a patent non-aggression clause; it is MIT licensed because the GPLv2 is not compatible with the Apache license. Other licenses apply because Rust includes source from other projects such as LLVM. None of this can be changed with any reasonable amount of effort.

**Comment by edunham** [comment]
Back to the topic of this issue, I agree that it's almost ready to be closed. Actions that other contributors can help us with to wrap up the loose ends are: 

* Find and PR some kind of open data license to crates.io-index (discuss this with @alexcrichton)
* PR a CC-BY license to the blog repo, and get sign-off from all contributors to that repo before merging (https://github.com/rust-lang/blog.rust-lang.org/graphs/contributors)
* Follow up on https://github.com/rust-lang/rfcs/issues/1259 -- research what other projects have done to sort out ambiguous copyright on standards, propose a suggestion to the core team.

**Comment by brson** [comment]
There's a thread open on irlo about the RFCs repo https://internals.rust-lang.org/t/relicense-the-rfcs-repo-under-the-cc-by-4-0-license/3870/3

**Comment by Mark-Simulacrum** [comment]
So the thread on IRLO went silent, and I don't think the RFCs repo ever got any form of license, among others. @brson If we still care about this (and we do, I think), we should probably have core team discuss and decide on a license for the RFCs repo and then move towards adding that.

**Comment by est31** [comment]
Opened an RFC with a proposal to settle the licensing situation of rust-lang/rfcs: https://github.com/rust-lang/rfcs/pull/2044

**Comment by est31** [comment]
The RFC got merged and got a tracking issue #43461 .

**Comment by apiraino** [comment]
By looking at the [checkbox list](https://github.com/rust-lang/rust/issues/25664#issuecomment-131996315), can we say that this issue is now complete - perhaps can be closed? thanks

**Comment by est31** [comment]
@apiraino since that comment, the github organization had new repos added to it. I went over the non-fork repos in the rust-lang organization and found the following repos without a top-level LICENSE file:

- [ ] https://github.com/rust-lang/rustc-perf has LICENSE files in some of the sub directories, but not all of them.
- [ ] https://github.com/rust-lang/docker-rust
- [ ] https://github.com/rust-lang/docker-rust-nightly
- [ ] https://github.com/rust-lang/crater
- [ ] https://github.com/rust-lang/bors
- [ ] https://github.com/rust-lang/surveys
- [ ] https://github.com/rust-lang/fmt-rfcs
- [ ] https://github.com/rust-lang/.github
- [ ] https://github.com/rust-lang/rust-lang.github.io

Plus a bunch of workgroup/team repos:

- [ ] https://github.com/rust-lang/project-rfc-2229
- [ ] https://github.com/rust-lang/wg-debugging (empty repository)
- [ ] https://github.com/rust-lang/wg-cargo-std-aware
- [ ] https://github.com/rust-lang/wg-allocators
- [ ] https://github.com/rust-lang/moderation-team
- [ ] https://github.com/rust-lang/wg-incr-comp
- [ ] https://github.com/rust-lang/compiler-team-prioritization
- [ ] https://github.com/rust-lang/rustc-timing

So this issue should very much not be closed, but thanks for pinging this issue because it motivated this mini-survey of mine.

**Comment by apiraino** [comment]
thanks @est31 for the important update :)

**Comment by sanmai-NL** [comment]
What about applying tooling that checks for such things as licensing and SPDX headers, etc.? Checking for top level licenses is rote work and does not cover every concern.

**Comment by sanmai-NL** [comment]
Even without a licensing tool for CI, you could [query the GitHub API](https://docs.github.com/en/rest/licenses).

**Comment by est31** [comment]
@sanmai-NL yeah it would be nice to have some kind of automatted checking of this, and maybe every time there is a new repo being added, it would file a report to a dedicated github issue (or something like that).

**Comment by Kobzol** [comment]
docker-rust has been [licensed](https://github.com/rust-lang/docker-rust/issues/207).
